### ISSUE TITLE
Adopt TorchAO int8 activation training for hardware-accelerated quantized training

### ISSUE THREAD

**Issue by Lev-Stambler** [DeepOpt-com/WrinkleFree-Monorepo#7]
## Summary

Proposal to adopt TorchAO's `int8_mixed_precision_training()` for hardware-accelerated int8 activation quantization, replacing our custom fake-quantization implementation. This would provide **~1.7x training throughput** by leveraging INT8 Tensor Cores.

### Current Implementation
Our `BitLinear` layer uses custom activation quantization:
- **Location**: `src/wrinklefree/quantization/activation_quant.py`, `src/wrinklefree/models/bitlinear.py`
- **Method**: Per-token absmax scaling (`scale = 127 / max(|X|)`)
- **Limitation**: Fake quantization only - no hardware acceleration

```python
# Current approach (BitLinear.activation_quant)
scale = 127.0 / x.abs().max(dim=-1, keepdim=True).values.clamp(min=self.eps)
x_quant = (x * scale).round().clamp(-128, 127) / scale
```

### TorchAO Int8 Training

TorchAO provides hardware-accelerated int8 training via `int8_mixed_precision_training()`:

| Feature | Current | TorchAO |
|---------|---------|---------|
| Activation quantization | Fake (BF16 compute) | Real INT8 Tensor Cores |
| Throughput | Baseline | **~1.7x faster** |
| Scaling method | Per-token absmax | Row-wise abs-max (equivalent) |
| Hardware utilization | No acceleration | Uses INT8 tensor cores |

**Benchmark** (from TorchAO docs, Llama2-1B single GPU):
- BF16 baseline: ~17k tokens/second
- INT8 mixed-precision: ~29k tokens/second

### Phase 1: Training Integration
```python
from torchao.prototype.quantized_training import int8_mixed_precision_training
from torchao import quantize_

# After model creation
model = create_bitnet_model(...)
quantize_(model, int8_mixed_precision_training())
```

**Key decision**: Keep ternary weight quantization (BitLinear) but replace activation quantization with TorchAO.

### Phase 2: Model Architecture Updates
- [ ] Modify `BitLinear` to use TorchAO's int8 activation path
- [ ] Keep custom ternary weight quantization (no TorchAO equivalent)
- [ ] Ensure STE gradient flow still works with TorchAO

### Phase 3: Inference Compatibility
- [ ] Export to HuggingFace with `TorchAoConfig`
- [ ] Test compatibility with vLLM, TGI
- [ ] Verify training-serving consistency (WYTIWYS)

```python
# HuggingFace export with TorchAO
from transformers import TorchAoConfig

quantization_config = TorchAoConfig("int8_dynamic_activation_int8_weight")
model.save_pretrained("output", quantization_config=quantization_config)
```

### Compatibility with BitNet
TorchAO's int8 training is additive - we keep our ternary weight quantization:
- **Weights**: Custom BitLinear ternary {-1, 0, 1} (unchanged)
- **Activations**: TorchAO int8 (new, hardware-accelerated)

### FSDP2 Support
TorchAO int8 training works with FSDP2 (our current distributed setup).

### Requirements
- PyTorch 2.4+ (we use 2.5+)
- `torchao>=0.5.0`

## Questions to Resolve

1. **Gradient flow**: Does TorchAO's int8 activation quantization work with our STE for weights?
2. **Lambda warmup**: Can we still do gradual quantization warmup with TorchAO?
3. **Kernel availability**: Are INT8 tensor cores available on A100/H100?

## References

- [TorchAO GitHub](https://github.com/pytorch/ao)
- [TorchAO Quantized Training README](https://github.com/pytorch/ao/tree/main/torchao/prototype/quantized_training)
- [TorchAO QAT Blog](https://pytorch.org/blog/quantization-aware-training/)
- [Int8 Mixed-Precision PR #748](https://github.com/pytorch/ao/pull/748)
